File: custom/rightImageData/processMatrices.m
bbetdata = importdata('bb23.txt', '\n');
bbhtdata = importdata('bb21.txt', '\n');
hcetdata = importdata('hc23.txt', '\n');


numCells = length(hcValues) / 16; 

hcHetArray = cell(numCells,1);
bbHhtArray = cell(numCells,1);
bbHetArray = cell(numCells,1);

j = 1;
for i = 1:16:length(hcValues)
    hcHetArray(j) = num2cell(reshape(hcValues(i:i+15), [4, 4]), [1 2]);
    j = j + 1;
end

j = 1;
for i = 1:32:length(bbValues)
    bbHhtArray(j) = num2cell(reshape(bbValues(i:i+15), [4, 4]), [1 2]);
    j = j + 1;
end

j = 1;
for i = 17:32:length(bbValues)
    bbHetArray(j) = num2cell(reshape(bbValues(i:i+15), [4, 4]), [1 2]);
    j = j + 1;
end

bbHhcArray = cell(numCells,1);
for i = 1:numCells
    bbHhcArray(i) = num2cell(bbHetArray{i} / hcHetArray{i}, [1 2]);
end

htHhcArray = cell(numCells,1);
for i = 1:numCells
    htHhcArray(i) = num2cell(bbHhtArray{i} \ bbHhcArray{i} , [1 2]);
end

baselineT = htHhcArray{5}(1:3, 4);

avgT = zeros(3,1);
avgR = zeros(3,3);
figure
hold on
skipped = 0
for i = 1:numCells
    t = htHhcArray{i}(1:3, 4);
    R = htHhcArray{i}(1:3, 1:3);
    if norm(t - baselineT) > .03
        skipped = skipped + 1;
        continue
    end
    avgT = avgT + t;
    avgR = avgR + logm(R);
    mArrow3(t, t + R(:,1) * .1, 'color', 'red', 'stemWidth', 0.0002); 
end

avgT = avgT / (numCells - skipped)
avgR = expm(avgR / (numCells - skipped))
mArrow3(avgT, avgT + avgR(:,1) * .1, 'color', 'blue', 'stemWidth', 0.0002); 

File: custom/rightImageData/processMatrices.py
import numpy as np

bbetData = np.loadtxt('bb23.txt', delimiter=',')
bbhtData = np.loadtxt('bb125.txt', delimiter=',')
hcetData = np.loadtxt('hc23.txt', delimiter=',')

bbetIds = bbetData[:, 0]
bbhtIds = bbhtData[:, 0]
hcetIds = hcetData[:, 0]

validIds = np.intersect1d(bbetIds, np.intersect1d(bbhtIds, hcetIds))
bbetValidIndices = np.in1d(bbetIds, validIds)
bbhtValidIndices = np.in1d(bbhtIds, validIds)
hcetValidIndices = np.in1d(hcetIds, validIds)

bbetData = bbetData[bbetValidIndices, 1:]
bbhtData = bbhtData[bbhtValidIndices, 1:]
hcetData = hcetData[hcetValidIndices, 1:]

numTransforms = len(validIds)
bbHets = np.reshape(bbetData.T, (4, 4, -1), 3)
bbHhts = np.reshape(bbhtData.T, (4, 4, -1), 3)
hcHets = np.reshape(hcetData.T, (4, 4, -1), 3)

bbHhcs = np.zeros([4, 4, numTransforms])
for i in xrange(numTransforms):
    bbHhcs[:, :, i] = np.dot(bbHets[:, :, i], np.linalg.inv(hcHets[:, :, i]))

htHhcs = np.zeros([4, 4, numTransforms])
for i in xrange(numTransforms):
    htHhcs[:, :, i] = np.dot(np.linalg.inv(bbHhts[:, :, i]), bbHhcs[:, :, i])

avgT = np.zeros([3, 1])
avgR = np.zeros([3, 3])

baselineT = htHhcs[0:3, 3, 4].reshape(3, 1)
skipped = 0
for i in xrange(numTransforms):
    t = htHhcs[0:3, 3, i].reshape(3, 1)
    if np.linalg.norm(baselineT - t) > .03:
        skipped += 1
        continue
    avgT += htHhcs[0:3, 3, i].reshape(3, 1)
    avgR += htHhcs[0:3, 0:3, i]

avgT = avgT / (numTransforms - skipped)
avgR = avgR / (numTransforms - skipped)

u, s, v = np.linalg.svd(avgR)
avgR = np.dot(u, v)

print avgT
print avgR

f = open('transformation.txt', 'w')
np.savetxt(f, avgT, newline='\n')
f.write('----------------------\n')
np.savetxt(f, avgR, newline='\n')
